### PR TITLE
Inject sidecar in PowerMax

### PR DESCRIPTION
**Pull request by medegw01**
# Description
This PR supports sidecar injection for csi-driver

# Issues

List the issues impacted by this PR:

| Issue ID |
| -------- |
|     #89 #92      |

# Checklist:

- [x] I have performed a self-review of my own changes.

### CODE REVIEW

**Comment by ioeicicdsvc** [comment]
@medegw01: Thank you for submitting your pull request!

Please make sure you have followed our contributing guidelines. We will review it as soon as possible.

------------------------------------------------------------------------------------------------------------------------
We want your feedback! If you have any questions or suggestions regarding our contributing process/workflow, please reach out to us at karavi@dell.com.

File: cmd/karavictl/cmd/inject.go [hunk 1]
```diff
@@ -56,8 +52,8 @@ func NewInjectCmd() *cobra.Command {
 	Examples:
 	# Inject into an existing vxflexos CSI driver 
 	kubectl get secrets,deployments,daemonsets -n vxflexos -o yaml \
-	  | karavictl inject --image-addr 10.0.0.1:5000/sidecar-proxy:latest --proxy-host 10.0.0.1 \
-	  | kubectl apply -f -`,
+	  | karavictl inje| kubectl apply -f -ct --image-addr 10.0.0.1:5000/sidecar-proxy:latest --proxy-host 10.0.0.1 \
```
Comment (tdawe) [inline cmd/karavictl/cmd/inject.go:55]: The documented usage of `karavictl inject` doesn't look correct here.

File: cmd/karavictl/cmd/inject.go [hunk 2]
```diff
@@ -135,7 +134,74 @@ func NewInjectCmd() *cobra.Command {
 	return injectCmd
 }
 
-func buildProxyContainer(imageAddr, proxyHost string, insecure bool) *corev1.Container {
+func rollout(b []byte, cmd *cobra.Command) {
```
Comment (tdawe) [inline cmd/karavictl/cmd/inject.go:137]: Is this function used (commented out above on line 125)? If not it can be removed.

**Review by gallacher — CHANGES_REQUESTED** [review]
Check for formatting is failing.  Please fix the formatting.

File: cmd/karavictl/cmd/inject.go [hunk 3]
```diff
@@ -125,10 +136,13 @@ func NewInjectCmd() *cobra.Command {
 		},
 	}
 
+	//karavictl inject --insecure=true --image-addr amaas-eos-mw1.cec.lab.emc.com:5046/sidecar-proxy:aaron --proxy-host lglap155.lss.emc.com --proxy-port powerflex=10000 --proxy-port powermax=102000
+
 	injectCmd.Flags().String("proxy-host", "", "Help message for proxy-host")
 	injectCmd.Flags().String("image-addr", "", "Help message for image-addr")
 	injectCmd.Flags().Bool("insecure", false, "Allow insecure connections from sidecar-proxy to proxy-server (default: false)")
 	injectCmd.Flags().String("root-certificate", "", "The root certificate file used by the proxy server")
+	injectCmd.Flags().StringSlice("proxy-port", []string{}, "proxy start port in the form <storageSystemType>=<startinPort>")
```
Comment (ianbibby) [inline cmd/karavictl/cmd/inject.go]: Typo "startinPort"

File: cmd/karavictl/cmd/inject.go [hunk 4]
```diff
@@ -186,15 +200,21 @@ func buildProxyContainer(ns, configName, imageAddr, proxyHost string, insecure b
 	return &proxyContainer
 }
 
+const (
+	DefaultStartingPortRange = 9000
```
Comment (ianbibby) [inline cmd/karavictl/cmd/inject.go:203]: Exported constant will need a comment.

File: cmd/karavictl/cmd/inject.go [hunk 5]
```diff
@@ -1024,3 +1044,60 @@ func scrubLoginCredentials(s []SecretData) []SecretData {
...
+		return map[string]int{
+			"powerflex": DefaultStartingPortRange,
+			"powermax":  DefaultStartingPortRange + 200,
+		}, nil
+	}
+
+	portRanges := make(map[string]int)
+	for _, v := range proxyPortFlags {
+		t := strings.Split(v, "=")
+		port, err := strconv.Atoi(t[1])
```
Comment (ianbibby) [inline cmd/karavictl/cmd/inject.go:1058]: Check `len(t)` in case it's 1, not 2.

**Review by ianbibby — CHANGES_REQUESTED** [review]

File: cmd/karavictl/cmd/inject.go (hunk 4 above)
Comment (atye) [reply cmd/karavictl/cmd/inject.go:203]: Should be fixed by:

https://github.com/dell/karavi-authorization/pull/86/commits/bd9df2a676bcd24109d79c81e6a3537938ce2780

File: cmd/karavictl/cmd/inject.go (hunk 3 above)
Comment (atye) [reply cmd/karavictl/cmd/inject.go]: Should be fixed by:

https://github.com/dell/karavi-authorization/pull/86/commits/4f914dca2ce92655ca2be27ea98aaae295b42962

File: cmd/karavictl/cmd/inject.go (hunk 5 above)
Comment (atye) [reply cmd/karavictl/cmd/inject.go:1061]: Updated.

File: docs/GETTING_STARTED_GUIDE.md [hunk 6]
```diff
@@ -364,7 +364,7 @@ Given a setup where Kubernetes, a storage system, a CSI driver, and Karavi Autho
     echo === Applying token token ===
     kubectl apply -f /tmp/token.yaml -n $DriverNameSpace
 
-    echo === injecting sidecar in CSI driver host === 
+    echo === injecting sidecar in all CSI driver host ===
```
Comment (gallacher) [inline docs/GETTING_STARTED_GUIDE.md:367]: hosts

File: cmd/karavictl/cmd/inject.go [hunk 7]
```diff
@@ -315,11 +383,45 @@ func (lc *ListChange) injectRootCertificate(rootCertificate string) {
...
+				if e.Name == "X_CSI_POWERMAX_ARRAYS" {
+					systemIDs = e.Value
+				}
+			}
+			break
+		}
+	}
+
+	if endpoint == "" || systemIDs == "" {
+		return nil, errors.New("could not fnd endpoint or system ID")
```
Comment (tdawe) [inline cmd/karavictl/cmd/inject.go:406]: Typo: fnd -> find

**Review by gallacher — DISMISSED** [review]

**Review by gallacher — DISMISSED** [review]
It seems there's quite a lack of logging going on here.  Maybe I'm missing something but I don't see many debug/info logs that might help diagnose issues in the future.  Other than that it looks good to me.

**Review by tdawe — DISMISSED** [review]

**Review by gallacher — APPROVED** [review]

**Review by tdawe — APPROVED** [review]

**Pull request merged**